cheetah/accelerator.py

```diff
@@ -299,6 +307,382 @@ def defining_features(self) -> list[str]:
+        cell_indices = torch.floor(normalized_pos).type(torch.long)
+
+        # Calculate the weights for all surrounding cells
+        offsets = torch.tensor([[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1]])
+        surrounding_indices = cell_indices.unsqueeze(1) + offsets  # Shape: (n_particles, 8, 3)
+        weights = 1 - torch.abs(normalized_pos.unsqueeze(1) - surrounding_indices)  # Shape: (n_particles, 8, 3)
+        cell_weights = weights.prod(dim=2)  # Shape: (n_particles, 8)
+
+        # Extract forces from the grids
+        def _get_force_values(force_grid):
+            idx_x, idx_y, idx_s = surrounding_indices.view(-1, 3).T
+            valid_mask = (idx_x >= 0) & (idx_x < grid_shape[0]) & \
+                        (idx_y >= 0) & (idx_y < grid_shape[1]) & \
+                        (idx_s >= 0) & (idx_s < grid_shape[2])
+            
+            valid_indices = torch.stack([idx_x[valid_mask], idx_y[valid_mask], idx_s[valid_mask]], dim=0)
+            force_values = force_grid[tuple(valid_indices)]
+            return force_values, valid_mask
+
+        Fx_values, valid_mask_x = _get_force_values(grad_x)
+        Fy_values, valid_mask_y = _get_force_values(grad_y)
+        Fz_values, valid_mask_z = _get_force_values(grad_z)
+
+        # Compute interpolated forces
+        interpolated_forces = torch.zeros((particle_pos.shape[0], 3))
+        values_x = cell_weights.view(-1)[valid_mask_x] * Fx_values * elementary_charge
+        values_y = cell_weights.view(-1)[valid_mask_y] * Fy_values * elementary_charge
+        values_z = cell_weights.view(-1)[valid_mask_z] * Fz_values * elementary_charge
+        
+        indices = torch.arange(particle_pos.shape[0]).repeat_interleave(8)
+        interpolated_forces.index_add_(0, indices[valid_mask_x], torch.stack([values_x, torch.zeros_like(values_x), torch.zeros_like(values_x)], dim=1))
+        interpolated_forces.index_add_(0, indices[valid_mask_y], torch.stack([torch.zeros_like(values_y), values_y, torch.zeros_like(values_y)], dim=1))
+        interpolated_forces.index_add_(0, indices[valid_mask_z], torch.stack([torch.zeros_like(values_z), torch.zeros_like(values_z), values_z], dim=1))
+
+        return interpolated_forces
+    
+
+    def track(self, incoming: ParticleBeam) -> ParticleBeam:
+        """
+        Track particles through the element. The input must be a `ParticleBeam`.
+        :param incoming: Beam of particles entering the element.
+        :return: Beam of particles exiting the element.
+        """
+        if incoming is Beam.empty or incoming.particles.shape[0] == 0:
+            return incoming
+        elif isinstance(incoming, ParticleBeam):
+            grid_dimensions = self._compute_grid_dimensions(incoming)
+            cell_size = 2*grid_dimensions / torch.tensor(self.grid_shape)
+            dt = self.length / (c*self._betaref(incoming))
```

Comment:
```suggestion
            dt = self.length / (c*self._betaref(incoming))
            # Copy the array of coordinates to avoid modifying the incoming beam
            outcoming_particles = torch.empty_like(incoming.particles)
            outcoming_particles[...] = incoming.particles
            outcoming = ParticleBeam(
                outcoming_particles,
                incoming.energy,
                particle_charges=incoming.particle_charges,
                device=particles.device,
                dtype=particles.dtype,
            )
```
And then pass `outcoming` to all the function that were taking `incoming` + `return outcoming`